rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1562 @@
+                client.clone(),
+                &command_options,
+            )),
+            update_asset_status_command_invoker: Arc::new(
+                adr_name_gen::UpdateAssetStatusCommandInvoker::new(
+                    application_context.clone(),
+                    client.clone(),
+                    &command_options,
+                ),
+            ),
+            notify_on_asset_update_command_invoker: Arc::new(
+                adr_name_gen::SetNotificationPreferenceForAssetUpdatesCommandInvoker::new(
+                    application_context,
+                    client,
+                    &command_options,
+                ),
+            ),
+            asset_update_notification_dispatcher,
+        })
+    }
+
+    /// Convenience function to get all observed device & inbound endpoint names to quickly unobserve all of them before cleaning up
+    #[must_use]
+    pub fn get_all_observed_device_endpoints(&self) -> Vec<(String, String)> {
+        let mut device_endpoints = Vec::new();
+        for device_receiver_id in self
+            .device_update_notification_dispatcher
+            .get_all_receiver_ids()
+        {
+            // best effort, if the id can't be parsed, then skip it
+            if let Some((device_name, inbound_endpoint_name)) =
+                Self::un_hash_device_endpoint(&device_receiver_id)
+            {
+                device_endpoints.push((device_name, inbound_endpoint_name));
+            }
+        }
+        device_endpoints
+    }
+
+    /// Convenience function to get all observed asset names to quickly unobserve all of them before cleaning up
+    #[must_use]
+    pub fn get_all_observed_assets(&self) -> Vec<(String, String, String)> {
+        let mut assets = Vec::new();
+        for receiver_id in self
+            .asset_update_notification_dispatcher
+            .get_all_receiver_ids()
+        {
+            // best effort, if the id can't be parsed, then skip it
+            if let Some((device_name, inbound_endpoint_name, asset_name)) =
+                Self::un_hash_device_endpoint_asset(&receiver_id)
+            {
+                assets.push((device_name, inbound_endpoint_name, asset_name));
+            }
+        }
+        assets
+    }
+
+    /// Shutdown the [`Client`]. Shuts down the underlying command invokers.
+    ///
+    /// Note: If this method is called, the [`Client`] should not be used again.
+    /// If the method returns an error, it may be called again to re-attempt unsubscribing.
+    ///
+    /// Returns Ok(()) on success, otherwise returns [`struct@Error`].
+    /// # Errors
+    /// [`struct@Error`] of kind [`ShutdownError`](ErrorKind::ShutdownError)
+    /// if any of the invoker unsubscribes fail or if the unsuback reason code doesn't indicate success.
+    /// This will be a vector of any shutdown errors, all invokers will attempt to be shutdown.
+    pub async fn shutdown(&self) -> Result<(), Error> {
```

Comment:
add test for this (should return all errors without network connectivity, but validates that it attempts all of them)

Reply:
for checking that 6 shutdown were called i couldnt find a way to do it without mocking invokers , otherwise it is just returning Ok

Reply:
the invokers should be created on new(), so they should be there - are they returning Ok() because they never subscribed, so we don't call unsubscribe and the shutdown for them succeeds since it doesn't need a network op?

Reply:
yes most likely